ce/amf/demands.ts

```diff
@@ -83,4 +132,195 @@ export class DemandBlock extends Entity {
+  }
+
+  const front = parts.slice(0, n).join('');
+  const back = parts.slice(n + 1).join('');
+
+  return parts[n].split(delimiter).filter(each => each).map(each => `${front}${each}${back}`).join(delimiter);
+}
+
+/** filters output and produces a sandbox context object */
+function filter(expression: string, content: string) {
+  const parsed = /^\/(.*)\/(\w*)$/.exec(expression);
+  const output = <any>{
+    $content: content
+  };
+  if (parsed) {
+    const filtered = new RegExp(parsed[1], parsed[2]).exec(content);
+
+    if (filtered) {
+      for (const [i, v] of filtered.entries()) {
+        if (i === 0) {
+          continue;
+        }
+        output[`$${i}`] = v;
+      }
+    }
+  }
+  return output;
+}
+
+export class Unless extends DemandBlock implements AlternativeFulfillment {
+
+  readonly from = new Strings(undefined, this, 'from');
+  readonly where = new Strings(undefined, this, 'where');
+
+  private data = <any>{};
+
+  get run(): string | undefined { return this.asString(this.getMember('run')); }
+  set run(value: string | undefined) { this.setMember('run', value); }
+
+  get select(): string | undefined { return this.asString(this.getMember('select')); }
+  set select(value: string | undefined) { this.setMember('select', value); }
+
+  get matches(): string | undefined { return this.asString(this.getMember('is')); }
+  set matches(value: string | undefined) { this.setMember('is', value); }
+
+  /** @internal */
+  override *validate(): Iterable<ValidationError> {
+    // todo: what other validations do we need?
+    yield* super.validate();
+    if (this.has('unless')) {
+      yield {
+        message: '"unless" is not supported in an unless block',
+        range: this.sourcePosition('unless'),
+        category: ErrorKind.InvalidDefinition
+      };
+    }
+  }
+
+  override async init(session: Session): Promise<Unless> {
+
+    if (this.usingAlternative === undefined) {
+      this.usingAlternative = false;
+      if (this.from.length > 0 && this.where.length > 0) {
+        // we're doing some kind of check.
+        const locations = [...this.from].map(each => expandEnvironment(session.environment, each).split(delimiter)).flat();
+        const binaries = [...this.where].map(each => expandEnvironment(session.environment, each));
+
+        const search = locations.map(location => binaries.map(binary => join(location, binary).replace(/\\/g, '/'))).flat();
+
+        // when we find an adequate match, we stop looking
+        // to do so and not work hrd
+
+        const Break = <NodeJS.ErrnoException>{};
+        for await (const item of stream(search, {
+          concurrency: 1,
+          stats: false, fs: <any>{
+            lstat: (path: string, callback: (error: NodeJS.ErrnoException | null, stats: Stats) => void) => {
+              // if we're done iterating, always return an error.
+              if (this.usingAlternative) {
+                return callback(Break, <Stats><any>undefined);
```

Comment:
What is this callback function?

Reply:
This is implementing the 'stat' function that the glob library needs -- the callback is the callback passed in from the library itself.